jena-fuseki2/jena-fuseki-access/src/main/java/org/apache/jena/fuseki/access/SecurityContext.java

```diff
@@ -78,11 +83,22 @@ public default Collection<String> visibleGraphNames() {
 
     public boolean visableDefaultGraph();
 
-    public default QueryExecution createQueryExecution(String queryString, DatasetGraph dsg) {
-        return createQueryExecution(QueryFactory.create(queryString), dsg);
+    public default QueryExecution createQueryExecution(HttpAction action, String queryString, DatasetGraph dsg) {
+        return createQueryExecution(action, QueryFactory.create(queryString), dsg);
     }
 
-    public QueryExecution createQueryExecution(Query query, DatasetGraph dsg);
+    public default QueryExecution createQueryExecution(HttpAction action, Query query, DatasetGraph dsg) {
+        QueryExecDatasetBuilder builder = QueryExec.newBuilder()
+            .dataset(dsg)
+            .query(query)
+            .context(action.getContext())
+            ;
+        // Note: Instead of this call (and the above build), the equivalent code from
```

Comment:
Repeating the same step here for timeout seems not right (but I wasn't sure what a better / the right approach would be).